runtime/extern.c

```diff
@@ -906,6 +1005,26 @@ static intnat extern_value(struct caml_extern_state* s, value v, value flags,
   extern_rec(s, v);
   /* Record end of output */
   close_extern_output(s);
+  /* Compress if requested */
+#ifdef HAS_ZSTD
+  if (s->extern_flags & COMPRESSED) {
+    uintnat uncompressed_len = extern_output_length(s);
+    extern_compress_output(s);
+    res_len = extern_output_length(s);
+    /* Write the header in compressed format */
+    store32(header, Intext_magic_number_compressed);
+    int pos = 5, len;
+    len = storevlq(header + pos, res_len); pos += len;
+    len = storevlq(header + pos, uncompressed_len); pos += len;
+    len = storevlq(header + pos, s->obj_counter); pos += len;
+    len = storevlq(header + pos, s->size_32); pos += len;
+    len = storevlq(header + pos, s->size_64); pos += len;
+    CAMLassert(len < 64);
```

Comment:
I don't understand the assertion here, and I am uneasy about the overflow behavior of the `storevlq` function.

At this point `len` is the number of base-128 digits of the last field read. The check `len < 64` on just this last length is odd.

At first I thought that maybe you meant to check that `pos < MAX_INTEXT_HEADER_SIZE`. But then this check would also be odd, because only runs that perform out-of-bound writes would fail the test -- so in particular the compiler may erase it.

I would feel more at ease if the `storevlq` and `readvlq` functions took as extra parameter the "end" of the input buffer, and abort if they reach the end while looking for the end of the digit.
(For `storevlq` this is arguably academic as we can statically bound the number of base-128 digits needed for an `uintnat` on today's machines.)

Reply:
The assertion was left over from an earlier design.  Currently, it is wrong (it should be about pos, not len) and useless (as you noticed), so I removed it in 754b1a2f9e.

For the foreseeable future, `uintnat` is at most 64 bits, so `storevlq` is guaranteed to write at most 10 bytes (since 9*7 < 64 <= 10*7).  I don't think we need any more bounds checking.

`readvlq` can be abused by artificially-padded input `0x80 ... 0x80 <the actual number>`.  However, the whole unmarshaler can be abused a zillion different ways by malicious input, so what the heck...

Besides, if we were to do bounds checking in `intern.c`, we'd have checks in `read8u` first, and then `readvlq` would automatically be checked.